dask/dataframe/io/parquet/fastparquet.py

```diff
@@ -433,14 +433,20 @@ def read_metadata(
             pf_piece = (file_path, gather_statistics) if pf_deps == "tuple" else pf_deps
             part_item = {
                 "piece": piece,
-                "kwargs": {"pf": pf_piece, "categories": categories_dict or categories},
+                "kwargs": {"pf": pf_piece},
             }
             parts.append(part_item)
 
         # Cannot allow `None` in columns if the user has specified index=False
         if index is False and None in meta.columns:
             meta.drop(columns=[None], inplace=True)
 
+        # Add `common_kwargs` to the first element of `parts`.
+        # We can return as a separate element in the future, but
+        # should avoid breaking the API for now.
+        if len(parts):
+            parts[0]["common_kwargs"] = {"categories": categories_dict or categories}
```

Comment:
So I would suggest that the ParquetFile without any row groups can be passed as common_kwargs
```
row_groups = pf.row_groups  # becomes the list of parts as above
pf.row_groups = []
pf.fmd.row_groups = []
```
Note that `ParquetFile._set_attrs` really does make unnecessary extra references to things which could be properties, hence the need to set empty lists twice. The individual row group objects serialise just fine with standard pickle, this is built into fastparquet.

Reply:
I agree that we should pass the "stripped-down" `ParquetFile` object in the `"common_kwargs"`.  However, since it will require a bit of "fastparquet"-engine re-organization (which could get slightly messy), I intend to do that in a separate PR.